Connect Frontend and Backend for Registration & Login

Teodora991 opened the pull request:
OVERVIEW This PR establishes complete integration between frontend and backend for user registration, email verification, and OTP-based login flows. All components are aligned with backend validation rules and follow consistent design patterns.

NEW COMPONENTS

Frontend:
HomePage: Authenticated user dashboard displaying email and logout button
CheckEmailPage: Post-registration page prompting users to verify email
VerificationSuccessPage: Success message after email verification
VerificationFailurePage: Failure message if verification link expired
AuthInterceptor: HTTP interceptor for JWT token injection

Backend:
MailTemplates: HTML email templates for account verification and OTP

CHANGES

Input Components:
Added min and max input validation to TextInput and PasswordInput
Applied constraints matching backend rules: Names: 2-20 characters Username: 4-20 characters Password: 10-20 characters

Dependencies: Added github.com/gorilla/handlers v1.5.2 for CORS support

Teodora991 committed:
Establish seamless integration between frontend and backend for
user registration, email verification, and OTP-based login workflows.

vanjmali commented on frontend/src/app/services/auth.service.ts:
```diff
@@ -1,144 +1,163 @@
-    });
-    // TODO: Send OTP via email service
-  }
+  private readonly API_BASE = 'http://localhost:3000/api/users'; // Traefik API Gateway on port 3000
```
Can this be loaded from environment variable?

vanjmali commented on user-service/services/user_service.go:
```diff
@@ -36,6 +37,8 @@ var (
 	// ErrBadCredentials indicates the credentials are invalid.
 	ErrBadCredentials = errors.New("invalid credentials")
+	// ErrEmailDelivery indicates the verification email could not be sent.
+	ErrEmailDelivery = errors.New("failed to send verification email")
```
This is never used, so this error will never appear. I assume you wanted to use it when it fails to send an email to respond with that error rather than 500?

vanjmali commented on .env.example:
```diff
@@ -40,9 +40,8 @@ SRV_USER_JWT_ACCESS_EXPIRATION_SECONDS=3600 # 1 hour
-SRV_USER_VERIFICATION_SUCCESS_URL=https://google.com
-SRV_USER_VERIFICATION_FAILURE_URL=https://apple.com
+SRV_USER_VERIFICATION_SUCCESS_URL=http://localhost:4200/verification-success
+SRV_USER_VERIFICATION_FAILURE_URL=http://localhost:4200/verification-failure
```
These should use the proxy http://localhost:3000 for local development.

Docker compose builds around port 3000 for all services, including frontend.

vanjmali commented on user-service/routers/router.go:
```diff
@@ -1,21 +1,35 @@
+		gorillaHandlers.AllowedOrigins([]string{"*"}),
+	)
+
+	return corsHandler(r)
```
Since the APIs are supposed to run from the same domain as the frontend app, no API requests would be cross-origin. We can review this if it's needed or not, otherwise should be moved to `common-lib` after discussion so it can be re-used in other microservices.

vanjmali commented on user-service/handlers/user_handler.go:
```diff
@@ -174,3 +179,23 @@ func (h *UserHandler) HandleVerifyLoginOtp(w http.ResponseWriter, r *http.Reques
+
+// HandleCheckEmail checks if an email is already registered
+func (h *UserHandler) HandleCheckEmail(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
```
This header set is handled by the `respond` package. It is not needed to be set.

vanjmali commented on user-service/handlers/user_handler.go:
```diff
@@ -174,3 +179,23 @@ func (h *UserHandler) HandleVerifyLoginOtp(w http.ResponseWriter, r *http.Reques
+	w.Header().Set("Content-Type", "application/json")
+
+	vars := mux.Vars(r)
+	email := vars["email"]
```
Why are we using this instead of a DTO struct?

vanjmali commented on user-service/validation/custom.go:
```diff
@@ -46,3 +46,15 @@ var CheckValidUsername = requests.CustomValidator{
+	}
+
+	return true
+}
```
We've introduced `requests.RegisterValidation` (to be used at `main.go`) for this to also have a friendly error message appear if field is invalid upon request body validation.

See [`validaiton.CheckValidUsername`](https://github.com/vanjmali/spotlite/blob/72563d72fce9399da57f24c16640df6310ee37ec/user-service/validation/custom.go#L31) for more information.

vanjmali commented on user-service/services/mail_service.go:
```diff
@@ -18,44 +17,64 @@ func InitMailingService(client *mail.Client) *MailService {
+	m.Subject("Verify your Spotlite account")
+
+	// Render email template with verification URL
+	verificationURL := "http://localhost:3000/api/users/verify?token=" + token
```
Either re-add the `// TODO` comment or replace the hardcoded URL with the environment vairable.

vanjmali commented on user-service/services/mail_service.go:
```diff
@@ -18,44 +17,64 @@ func InitMailingService(client *mail.Client) *MailService {
-func (ms *MailService) sendAccountVerificationEmail(mailto string, token string) {
+func (ms *MailService) sendAccountVerificationEmail(mailto string, token string) error {
 	m := mail.NewMsg()
 	if err := m.From("mail@spotlite.com"); err != nil {
```
The FROM header should also be configurable via environment varaible. Either add a TODO comment or environemnt variable. Don't forget we have `utils` package within `common-lib` for reading environment variables.

vanjmali commented on user-service/services/mail_service.go:
```diff
@@ -18,44 +17,64 @@ func InitMailingService(client *mail.Client) *MailService {
 func (ms *MailService) SendLoginOtp(mailto string, otp string) error {
 	m := mail.NewMsg()
 
 	if err := m.From("mail@spotlite.com"); err != nil {
```
The FROM header should also be configurable via environment varaible. Either add a TODO comment or environemnt variable. Don't forget we have `utils` package within `common-lib` for reading environment variables.

vanjmali reviewed (changes requested):
I haven't yet taken a look at the frontend look, but did review the backend for now.

Teodora991 committed:
Resolve PR comments.

stevanovicm5 commented:
When i try to paste OTP code, it only goes on one field. I think we should change it so it can be pasted all at once. It's little annoying to type all 6 digits.

stevanovicm5 commented on user-service/services/mail_service.go:
```diff
@@ -18,44 +24,64 @@ func InitMailingService(client *mail.Client) *MailService {
-	if err := ms.c.DialAndSend(m); err != nil {
-		log.Fatalf("failed to send mail: %s", err)
+	// Render email template with verification URL
+	verificationURL := VerificationSuccessUrl + "?token=" + token
```
When i register and click on link sent on my email, user stays `INACTIVE`, it only renders the front page.
I think it's because you need to forward it to backend first `/verify` which then redirects to the front page.

stevanovicm5 replied:
resolved in #23

stevanovicm5 commented on user-service/routers/router.go:
```diff
@@ -1,21 +1,27 @@
 
 // HandleRequests wires HTTP routes to user handlers.
-func HandleRequests(h *handlers.UserHandler, rth *handlers.RefreshTokenHandler) *mux.Router {
+func HandleRequests(h *handlers.UserHandler, rth *handlers.RefreshTokenHandler) http.Handler {
```
why did we change it to http.Handler?

Teodora991 replied:
Change was made to better align with Go conventions.
This keeps the API more flexible by returning an interface instead of a concrete type, makes it easier to wrap with middleware or change the underlying implementation later, and is generally more idiomatic Go.
Since *mux.Router already implements http.Handler, this doesn’t introduce any breaking changes — it just improves the overall design.

stevanovicm5 commented on user-service/go.mod:
```diff
@@ -16,6 +16,7 @@ require (
 replace github.com/vanjmali/spotlite/common-lib => ../common-lib
 
 require (
+	github.com/felixge/httpsnoop v1.0.3 // indirect
```
this is not used

Teodora991 replied:
Nice catch. I added the change to the last PR ( feature/home-page-with-header)

Teodora991 committed:
Implement OTP resend logic.  (#22)

NOTE: The folder structure has changed. Please ignore changes in files
under shared/components/input, as they were only relocated from
shared/input to shared/components/input.

stevanovicm5 committed:
Merge remote-tracking branch 'origin/develop' into feature/auth-fe-be-alignment

stevanovicm5 committed:
fix: use common lib
